### PR TITLE
Improved "include" fix performance: optimized normalizePath method

### LINKED ISSUE #1856: Optimize updates
Based on some profiling data we did for a client, it seemed that Liquibase was recompiling a lot of often used regexes, e.g.:
image

The latest code also seems to contain lots of them, e.g. https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/changelog/DatabaseChangeLog.java#L645-L650, resulting in a horrible performance.

Extracting every regex on the critical path shouldn't be difficult. Creating fewer Strings may be more difficult, but would be worth it.
I'd gladly help, but I don't have the reproducer code available, our client does.

### LINKED ISSUE #2221: normalizePath super slow?
## Environment

**Liquibase Version**: 4.3.5
**Liquibase Integration & Version**: spring boot
**Database Vendor & Version**: mysql
**Operating System Type & Version**: linux

## Description

Super slow run after liquibase upgrade. It takes 5~10 minutes to run. In previous versiones it took < 30 seconds.

## Steps To Reproduce

I have a change set xml file with more than 500 xml includes, like this:

`<include file="another.xml" relativeToChangelogFile="true" />`

## Actual Behavior
Takes a lot of time to run command

## Expected/Desired Behavior
Similar run time than previous versions

## Screenshots (if appropriate)
I've run with profiler, and got this:

screenshot2021-11-17T11:51:46-03:00

liquibase.changelog.DatabaseChangeLog#normalizePath
this method takes a lot of time.
i suspect that precompiling the regexp will help.

## Additional Context
if it looks good, i can provide a patch.

### PR DESCRIPTION
**Pull request by lzxgyh**
## Impact

- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [x] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

Replaces on-the-fly regular expression usage with pre-compiled versions.

Fixes #2221 
Fixes #1856

## Things to be aware of

- Nothing

## Things to worry about

- Nothing

### CODE REVIEW

**Comment by nvoxland** [comment]
Thanks @lzxgyh!
I updated the description to remove the boilerplate template with what was done.

I worry about the creation of the matcher static variables, though. If this gets used in a multithreaded way, I think those will cause problems. If you just create the matchers as local variables instead of resetting the static variables does it loose a lot of the performance gains? The original "compile" call should be what makes the most difference, right?

**Comment by lzxgyh** [comment]
> Thanks @lzxgyh! I updated the description to remove the boilerplate template with what was done.
> 
> I worry about the creation of the matcher static variables, though. If this gets used in a multithreaded way, I think those will cause problems. If you just create the matchers as local variables instead of resetting the static variables does it loose a lot of the performance gains? The original "compile" call should be what makes the most difference, right?

**Comment by lzxgyh** [comment]
> Thanks @lzxgyh! I updated the description to remove the boilerplate template with what was done.
> 
> I worry about the creation of the matcher static variables, though. If this gets used in a multithreaded way, I think those will cause problems. If you just create the matchers as local variables instead of resetting the static variables does it loose a lot of the performance gains? The original "compile" call should be what makes the most difference, right?

Hi  [nvoxland](https://github.com/nvoxland)

I think you are right, in the past, I always thought that liquibase was a single-threaded tool

**Comment by lzxgyh** [comment]
> Thanks @lzxgyh! I updated the description to remove the boilerplate template with what was done.
> 
> I worry about the creation of the matcher static variables, though. If this gets used in a multithreaded way, I think those will cause problems. If you just create the matchers as local variables instead of resetting the static variables does it loose a lot of the performance gains? The original "compile" call should be what makes the most difference, right?

Hi  [nvoxland](https://github.com/nvoxland)

I think you are right, in the past, I always thought that liquibase was a single-threaded tool

**Comment by nvoxland** [comment]
Are you able to fix that up quick, @lzxgyh?

**Comment by lzxgyh** [comment]
Sure, I fpushed a new commit to remove the matcher static variables , Please review @nvoxland

**Review by nvoxland — APPROVED** [review]
Code review and test results:

Things to be aware of:
- Just moves implementation of where the regular expressions are managed from one spot to another, no functionality change
- I did not re-test the performance improvements, but I can see from the performance dumps that this what a large problem in the code and the PR will fix it as the contributor says it does

Things to worry about:
- Nothing

File: liquibase-core/src/main/java/liquibase/changelog/DatabaseChangeLog.java [hunk 1]
```diff
@@ -725,14 +733,12 @@ public static String normalizePath(String filePath) {
...
         }
-        return filePath.replaceFirst("^classpath:", "")
-                .replaceAll("\\\\", "/")
-                .replaceAll("//+", "/")
-                .replaceAll("/\\./", "/")
-                .replaceFirst("^[a-zA-Z]:", "")
-                .replaceFirst("^\\.?/", "")
-                ;
-
+        String noClassPathReplaced = NON_CLASSPATH_PATTERN.matcher(filePath).replaceFirst("");
```
Comment (XDelphiGrl) [inline liquibase-core/src/main/java/liquibase/changelog/DatabaseChangeLog.java:736]: Thank you, @lzxgyh - This is a nice improvement to code readability!

**Review by XDelphiGrl — APPROVED** [review]
* Fix improves performance of normalizePath.
* Fix improves readability of REGEX matching in DatabaseChangelog.java
* No additional testing necessary.

APPROVED

[Passing Functional Tests](https://github.com/liquibase/liquibase-test-harness/actions/runs/2702642521)
[Passing Test Harness Execution](https://github.com/liquibase/liquibase-test-harness/actions/runs/2702642521)

**Review by suryaaki2 — APPROVED** [review]

**Pull request merged**